### ISSUE TITLE
a Morepath version of the Django tutorial

### ISSUE THREAD

**Issue by faassen** [morepath/morepath#507]
The Django tutorial is a good introduction to Django: https://docs.djangoproject.com/en/1.10/intro/tutorial01/

It would be interesting to rewrite this into a Morepath version. The primary reason is to have the missing Morepath tutorial. We'd focus on server HTML.

The other reason is to identify whether Morepath is missing something we'd want. Something like the admin UI is an obvious big one. A less obvious one is support for CSRF injection prevention, though it could possibly take quite a different shape.

Yet another reason is to offer a nice way to compare Django and Morepath.

**Comment by href** [comment]
The question is how much we can replicate Django in this sense. Django is very much a batteries-included approach to web-application development. As a micro-framework we're never going to play in that league. Things we identify as lacking would be lacking in our ecosystem, not the framework itself - unless they are not achievable.

That being said, I do agree that Django's beginner-friendliness is something to be emulated. I'm sure it's one of the biggest reason for its success.

Our focus probably has to be different. Much like Pyramid or Flask, Morepath is always going to be harder to learn as there are no tightly integrated components. The best we can do is integrate well-chosen components and use them in a tutorial.

**Comment by faassen** [comment]
Yes, we'd certainly need to make choices in such a tutorial, like picking jinja2 as template language, SQLAlchemy for the database, a static file publishing system (I'm leaning against the whole bowerstatic stuff I came up with as too complicated), an authentication system (maybe, not sure yet), and so on. But showing how you can make such choices and how to integrate them is worthwhile in a tutorial by itself.

And yes, things lacking don't need to go into Morepath itself, but they can go into the documentation and the ecosystem. It's okay if the documentation makes some choices the framework doesn't, as long as we're clear that those are choices.

**Comment by henri-hulski** [comment]
I just rewrote morepath_sqlalchemy for Pony ORM. It looks so much simpler and don't need any wrapper. 
Just adding db_session to request. 
So I would recommend that. 

BTW more.transaction seems to be broken with zope.transaction 2.x.

**Comment by faassen** [comment]
Do you have a link to the Pony rewrite? I'm very interested in taking a look at it, and also how transaction integration works. You're saying it doesn't involve `zope.transaction` at all?

@henri-hulksi could you leave an issue for more.transaction? I think you're referring to `transaction` 2.x, not `zope.transaction`, correct? (I made the same mistake before

**Comment by henri-hulski** [comment]
@faassen Yeah you are right. 
I just have to fix the tests for morepath_ponyorm. It's just the same as morepath_sqlalchemy for easy comparison. 
If you're ok I can upload it to Morepath when it's finished.

**Comment by href** [comment]
> Could you leave an issue for more.transaction? I think you're referring to transaction 2.x, not zope.transaction, correct? (I made the same mistake before myself)

I made it last time we were discussing it 😓

> So I would recommend that.

Once again I would like to note that if you do things without a general transaction mechanism then you're going to have a bad time once your application becomes bigger and needs to tie in other things with your database transaction. This is just a matter of time.

Therefore we should really focus on tying everything up with more.transaction. If every more.* package does its own transaction handling we'll end up with many incompatible packages. This is why other frameworks like Pyramid went with the transaction package as well. To me more.transaction should be the canonical way.

But this is a bit off-topic.

**Comment by faassen** [comment]
Oh, sorry about that issue, I forgot about it/missed it.

Concerning transaction, it does have benefits, as it does retry out of the box, and it allows the integration of the transaction mechanisms of multiple databases, such as ZODB and relational. I did that with Grok in the past.

**Comment by henri-hulski** [comment]
Pony ORM has his own transaction manager and [db_session](https://docs.ponyorm.com/api_reference.html#transactions-db-session) also has a retry parameter.

**Comment by j-m-r** [comment]
On Dec 8, 2016, at 7:55 AM, Martijn Faassen <notifications@github.com> wrote:

> And yes, things lacking don't need to go into Morepath itself, but they can go into the documentation and the ecosystem. It's okay if the documentation makes some choices the framework doesn't, as long as we're clear that those are choices.

<uncloaking>

I strongly agree — I picked morepath specifically because it offers choices — but at the same time, it will be more attractive if there are worked-out examples that show how to integrate some instance of the "missing" parts.

**Comment by henri-hulski** [comment]
Continueing the ORM discussion on #485.

**Comment by yeraydiazdiaz** [comment]
I tend to agree with @href that Morepath is more akin to Flask that it is to Django, and given the recent rewrite of [Miguel Grinberg's Flask Mega-Tutorial](https://blog.miguelgrinberg.com/post/the-flask-mega-tutorial-part-i-hello-world) I thought it would be a good exercise to try and [recreate it in Morepath](https://github.com/yeraydiazdiaz/morepath_microblog).

Not sure if that aligns with the original intention of this issue, if so we can discuss how best to approach it.
